woocommerce/woocommerce
# Try: deprecate product types in favor of traits

### Changes proposed in this Pull Request:

This PR is a POC demonstrating one option for deprecating product types in favor of traits with a more formalized compatibility system.  It builds on the work in https://github.com/woocommerce/woocommerce/pull/44585.  This PR:

* Converts the variable product type to a trait
* Moves all functions related to that trait to its trait class
* Allows product methods to be overridden by traits.  All prefixed methods with `_` can be overridden by the trait if the trait is enabled on the product
* Allows traits to be enabled/disabled and easily handled via JSON responses
* Returns the correct type when `$product->get_type()` is called when the variations trait is enabled

While this seems to work in a very simple, contrived scenario, this concept is not without its drawbacks and has some still unresolved issues:

* The use of `__call()` to override product methods at runtime is pretty hacky.  This is primarily to maintain backwards compatibility with the existing methods, but we should probably consider alternatives to this if we do remove types.
* Type classes often make use of private or protected methods within the product, which will cause errors unless we explicitly allow them to pass through from traits (as done with `get_prop` in this PR).
* Trait method overrides can conflict.  Currently the first active trait with an override method (e.g., `_add_to_cart_text`) will be chosen to override the product class's method.  This works as is since the types being converted to traits aren't yet compatible with one another anyway and should not be active at the same time.  That said, it's not ideal long-term with multiple traits affecting a given area and might suggest a filter is more appropriate.
* There's not a 1:1 with existing traits in the product schema.  E.g., `downloadable` was at the top level and now exists under `traits.downloadable`.  This is because we need a generic array for `traits` so we can update via `WC_Data` and so that the client can more easily parse active traits.
* We'll need to decide whether enabling a trait that's not compatible with others a) disables the incompatible traits or b) throws a warning and prevents activation of said trait



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Programmatically create a product and check it's type
```php
$product = new \WC_Product();
echo $product->get_type(); // Should be `simple`
```
2. Set the product trait and check it's type to confirm the trait shows the deprecated type
```php
$product->enable_trait( 'variable' );
echo $product->get_type(); // Should be `variable`
```
3. Check that when the trait is enabled, we get its overridden methods.
```php
$product->enable_trait( 'variable' );
echo $product->add_to_cart_description(); // Select options for ...
```
4. Check that when the trait is disable, we get its overridden methods.
```php
$product->disable_trait( 'variable' );
echo $product->add_to_cart_description(); // Read more about...
```
5. Check that trait-specific methods can still be accessed when a trait is not enabled
```php
var_dump( $product->is_downloadable() );
```



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
